### PR TITLE
c-s CLI: Options parsing

### PR DESCRIPTION
**Pull request by muzarski**
Introduced modules:
* settings::option - Parses options and its parameters

Supported options:
-node
-schema
-rate

These options should be enough for implementing the runtime.

Apart from that - extended parameters' parsing module by introducing `MultiParam`. A multiparameter is a complex parameter that contains subparameters. For example, given the output of `./cassandra-stress help -schema`:
```
Usage: -schema [replication(?)] [keyspace=?] [compaction(?)] [compression=?]

  replication([strategy=?][factor=?][<option 1..N>=?]): Define the replication strategy and any parameters
      strategy=? (default=org.apache.cassandra.locator.SimpleStrategy) The replication strategy to use
      factor=? (default=1)                     The number of replicas
  keyspace=? (default=keyspace1)           The keyspace name to use
  compaction([strategy=?][<option 1..N>=?]): Define the compaction strategy and any parameters
      strategy=?                               The compaction strategy to use
  compression=?                            Specify the compression to use for sstable, default:no compression

```

`-schema` option uses two multiparameters: `replication` and `compaction`. They both have some predefined subparameters. `replication` accepts two of them, which are: `strategy`, and `factor`. Apart from that, both `replication` and `strategy` accept some arbitrary parameters (denoted by `<option 1..N>=?`). The arbitrary parameters must match pattern `key=value` and in the case of `-schema` option, will be correspondingly appended when defining `replication` and `compaction`.

`MultiParam` is responsible for parsing such parameters.

### CODE REVIEW

**Comment by avelanarius** [comment]
Small typo in PR description:
> A multiparameter is a complex parameter that contain**t** subparameters

**Comment by avelanarius** [comment]
This isn't maybe that important in this case (since this PR isn't that large), but it could have been split into smaller commits (probably one that adds MultiParam and one that adds all those node/rate/schema options).

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+use regex::Regex;
+
+use super::{Param, ParamCell, ParamHandle, ParamMatchResult};
+
+lazy_static! {
+    // Multiparameters must match the pattern:
+    // param_name(key1=value1,somethingThatMatchesPredefinedSubparam,key2=value2,...keyN=valueN)
+    // These regular expressions help to validate and extract the subparameters from the string.
+    static ref SUBPARAM: Regex = Regex::new(r"([^,]+)").unwrap();
+    static ref ARBITRARY_PARAM: Regex = Regex::new(r"^([^=]+)=([^=]+)$").unwrap();
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:13]: Just a question: why is `ARBITRARY_PARAM` regex a `^ ... $` regex while `SUBPARAM` is not?

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        for (_, [subparam]) in SUBPARAM.captures_iter(arg_val).map(|m| m.extract()) {
+            // Check if the argument matches on of the predefined subparameters.
+            match self.try_parse_predefined(subparam) {
+                ParamMatchResult::Error(e) => return Err(e),
+                ParamMatchResult::Match => continue,
+                _ => (),
+            }
+
+            // If any of the subparamters didn't match, try to parse as arbitrary parameter (if applicable).
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:141]:
```suggestion
                ParamMatchResult::NoMatch => (),
            }

            // If the subparameter didn't match, try to parse it as arbitrary parameter (if applicable).
```

I think it's clearer written this way - please correct me if I misunderstood the code.

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    fn set_satisfied(&mut self) {
+        self.satisfied = true;
+        for param in self.subparams.iter() {
+            param.borrow_mut().set_satisfied();
+        }
+        // Clear the subparameters so the user can consume them via corresponding handles.
+        // Otherwise, .get() method called on their handles would cause panic.
+        self.subparams.clear();
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:170]: "Clear" in the comment isn't descriptive enough to fully explain that by removing `ParamCell`s those values can then be safely used (it isn't explained that `get()` refers to `RefCell::get()`).

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs (hunk 1 above)
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:13]: To give an example:
Assume we receive `foo=bar,key=value,factor=3`. Firstly, the `SUBPARAM` regexp is used to iterate over substrings that match this pattern i.e. `foo=bar`, `key=value` and `something=something`. Note that we cannot wrap the regexp with `^$`, because we want to access multiple substrings and iterate over them (if we contained `^` and `$` we would get no match - so nothing to iterate over).

Now that we extracted three subparameters, we firstly try to parse them as some predefined subparameter (in `replication`'s case it's for example `factor` subparam). If it didn't match any of the prefedined subparameters, we try to parse it as an arbitrary subparameter (something that exactly matches `ARBITRARY_PARAM` pattern).

I hope this answers your question.

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs (hunk 2 above)
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:141]: I agree

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs (hunk 1 above)
Comment (avelanarius) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:13]: Yes, it makes sense. If I understand it right, your solution permits multiple ','? Like: `replication(foo=bar,,,,key=value)`. It might have been easier to just split the string on ',' instead of using the regex.

Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:13]: Now that I think of that, I should add some comment/more descriptive name for `SUBPARAM`. At first glance, it's not clear that it's used to extract the sub parameters delimited by a comma.

Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:13]: > Yes, it makes sense. If I understand it right, your solution permits multiple `,`? Like: `replication(foo=bar,,,,key=value)`. It might have been easier to just split the string on `,` instead of using the regex.

Oh, you are right. And yes, the split would make more sense here.

**Comment by wprzytula** [comment]
> This isn't maybe that important in this case (since this PR isn't that large), but it could have been split into smaller commits (probably: one that adds MultiParam and one that adds all those node/rate/schema options).

For me, splitting the changes into smaller commits is crucial for review performance. It's really hard to follow the changes in so many files at once.

**Comment by wprzytula** [comment]
> > This isn't maybe that important in this case (since this PR isn't that large), but it could have been split into smaller commits (probably: one that adds MultiParam and one that adds all those node/rate/schema options).
> 
> For me, splitting the changes into smaller commits is crucial for review performance. It's really hard to follow the changes in so many files at once.

Will you do it for the sake of reviewing this PR, @muzarski ?

**Comment by wprzytula** [comment]
PRs should group the content by feature. Here, the feature you introduce is _option parsing_. Commits, on the other hand, should be possibly small, clean steps that consitute a road towards that feature.

For reference, you can take a look at e.g. this: https://github.com/scylladb/scylla-rust-driver/pull/758/commits. That's how I structure my PRs. There are both bigger commits (as long as splitting them would be problematic, because their nature is monolythic) and commits as small as renaming or deleting one function. A rule of thumb is that if you can distil, extract a step in a way that it is quite independent and makes sense on its own, then you should make it a separate commit.

Keep in mind, however, that this is only my personal view and others might argue here.

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs [hunk 4]
```diff
@@ -134,18 +140,46 @@ where
...
+        let rate = RateOption::parse(&mut payload)?;
+        let schema = SchemaOption::parse(&mut payload)?;
+
+        // List the unknown options along with their parameters.
+        let build_unknown_arguments_err_message = || -> String {
+            let unknowns = payload
+                .iter()
+                .map(|(option, params)| {
+                    let params_str = params.join(" ");
+                    (*option).to_owned() + " " + &params_str
```
Comment (avelanarius) [inline src/bin/cql-stress-cassandra-stress/settings/mod.rs:161]: Nit: `format!()` would have been neater.

**Comment by avelanarius** [comment]
Nit: the commit titles should not end with `.` and the part after `:` should not start with a capital letter. You can be more specific in some commit titles with regards to the "module"/file it changes (as opposed to generic "c-s CLI")

**Comment by muzarski** [comment]
v2:
- Replaced string concatenation in `settings::mod.rs` with `format!`.
- Added simple test for parsing arbitrary parameters of multiparameters
- Changed commit descriptions.

**Comment by avelanarius** [comment]
Some feedback for commit message style:
- I tend to stick to 50/72 rule (commit title up to 50 characters, commit message body 72 character wrapping)
- As to commit message body, I tend to write them this way: "[Situation before the change] [Description of what changed] [Situation after the change] [Other info...]" (where it makes sense) and in an imperative mood. Some more resources about it: https://www.kernel.org/doc/html/v4.10/process/submitting-patches.html#describe-your-changes
  For example I'd rewrite:
  > c-s param: get rid of panic 
  > - Changed `Param` interface, so `parse` method returns `anyhow::Result`.
  > - Don't `panic` in case of error. Returning `anyhow::Err` instead.
  
  to:
  > c-s param: return Result instead of panicking
  > Before the change, `Param::parse` would panic if the Param
  > couldn't be parsed (`ParamMatchResult::NoMatch`, `ParamMatchResult::Error`).
  > A user of this method was supposed to validate it beforehand with
  > `try_match`.
  > Change `parse` to return a `Result<()>` instead of panicking. [...and some motivation for it...]

**Review by wprzytula — CHANGES_REQUESTED** [review]
I really liked how the PR was divided into commits. Thanks to that, the review went smoothly!

File: src/bin/cql-stress-cassandra-stress/settings/option/node.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+fn prepare_parser() -> (ParamsParser, NodeParamHandles) {
+    let mut parser = ParamsParser::new(NodeOption::CLI_STRING);
+
+    let datacenter = parser.simple_param(
+        "datacenter=",
+        r"^.*$",
+        None,
+        "Datacenter used for DCAwareRoundRobinLoadPolicy",
+        false,
+    );
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/node.rs:62]: Yes, I took it directly from c-s help message. But I 100% agree with you - let's adjust it to the Rust driver's naming.

File: src/bin/cql-stress-cassandra-stress/settings/option/node.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        "",
+        r"^[^=,]+(,[^=,]+)*$",
+        Some("localhost"),
+        "comma delimited list of nodes",
+        false,
+    );
+
+    // $ ./cassandra-stress help -node
+    // Usage: -node [datacenter=?] [whitelist] [file=?] []
+    parser.group(vec![&datacenter, &whitelist, &file, &nodes]);
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/node.rs:81]: Good point. I'm not used to using slices wherever possible yet.

[the remark this answers is not in the document]
Comment (wprzytula) [reply src/bin/cql-stress-cassandra-stress/settings/option/node.rs:81]: (or even `impl IntoIterator`)

File: src/bin/cql-stress-cassandra-stress/settings/option/rate.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        println!("  Auto: {}", self.auto);
+        if self.min_threads.is_some() {
+            println!("  Min threads: {}", self.min_threads.as_ref().unwrap());
+            println!("  Max threads: {}", self.max_threads.as_ref().unwrap());
+        } else {
+            println!("  Thread count: {}", self.threads.as_ref().unwrap());
+            println!("  OpsPer Sec: {}", self.throttle.as_ref().unwrap());
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/rate.rs:46]: > I can see the following assumption here: Either `min_threads` and `max_threads` are `Some`, or `threads` and `throttle` are `Some`.

You are right, that's why we "group" the parameters this way:
```
  // $ ./cassandra-stress help -rate
  // Usage: -rate threads=? [throttle=?] [fixed=?]
  //  OR
  // Usage: -rate [threads>=?] [threads<=?] [auto]
  parser.group(vec![&threads, &throttle, &fixed]);
  parser.group(vec![&threads_gte, &threads_lte, &auto]);
```

I agree that introducing an enum here would be better. I even recognized this issue when implementing the runtime - a lot of `if` statements when single `match` would do the job.

File: src/bin/cql-stress-cassandra-stress/settings/option/rate.rs [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    pub throttle: SimpleParamHandle,
+    pub fixed: SimpleParamHandle,
+    pub threads_gte: SimpleParamHandle,
+    pub threads_lte: SimpleParamHandle,
+    pub auto: SimpleParamHandle,
+}
+
+fn parse_per_second(arg: &str) -> u64 {
+    arg[..arg.len() - 2].parse::<u64>().unwrap()
+}
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/rate.rs:59]: Yes, that's something @piodul mentioned in the last PR. Currently, `ParamsParser` only VALIDATES the input and the parameters return the `String` that still needs to be PARSED. 

In this case, the parser (or rather the validator :)) makes sure that the CLI argument corresponding to this parameter will not panic when passed to the `parse_per_second` function.

piodul mentioned that I should follow the `parse, don't validate` rule. That's why I have a ready commit that introduces the parsing logic refactoring. Take a look if you are interested: https://github.com/muzarski/cql-stress/commit/38df015c2b4467817130bcb3582e264553ebb188. In short - the parameters are parameterized by the type that implements `Parsable` trait.

We decided to move on with what we have now, so it allows us to implement the actual runtime ASAP. But I still have in mind that the parsing logic should be refactored a bit. I'll take care of it once we merge the simple `write` workload.

File: src/bin/cql-stress-cassandra-stress/settings/option/rate.rs [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+    RateOption {
+        threads,
+        throttle,
+        fixed,
+        min_threads,
+        max_threads,
+        auto,
+    }
+}
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/rate.rs:143]: > This is strange that a parsing function does not return a `Result`. Can't it fail any case? `parse_per_second()` `unwrap()`s - maybe it should return a `Result` and `parse_with_handles` should propagate it?

As mentioned in this comment: https://github.com/scylladb/cql-stress/pull/42#discussion_r1286904434. The parser (`settings::param::parser::ParamsParser`) makes sure that `parse_per_second` won't panic. The parser's logic will be refactored in the future.

[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/rate.rs:143]: > I'd make this an associated function on `RateOption`.

100% agree - I have no idea why I left it this way.

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    fn try_parse_arbitrary(&mut self, arg: &str) -> Result<()> {
+        // Ensure that argument matches pattern "key=value"
+        anyhow::ensure!(
+            ARBITRARY_PARAM.is_match(arg),
+            "Invalid {} specification: {}",
+            self.prefix,
+            arg
+        );
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:103]: The error message is taken directly from c-s.
```
$ ./cassandra-stress write n=10 -schema 'replication(foo==bar)'
Invalid replication specification: foo==bar
...
... help message
```

Would you suggest any other error message we could use here?

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+        let split = arg.split('=').collect::<Vec<&str>>();
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:105]: True, It's yet another time I collect to vector (for no reason) instead of working with iterator directly. Need to get used to it...

File: src/bin/cql-stress-cassandra-stress/settings/command/help.rs [hunk 12]
```diff
@@ -40,5 +46,6 @@ pub fn parse_help_command(payload: &mut ParsePayload) -> Result<CommandParams> {
         return print_command_help(params[0]);
     }
 
-    todo!("Implement help for options.");
+    // Print help for option.
+    print_option_help(payload.iter().next().unwrap().0)
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/help.rs:50]: > The one `unwrap()` that still begins the function is upsetting. Please consider the possible ways to avoid it (my quick research hints that it can be too expensive), considering the caller's logic)

That's true. The fact is that the caller of this function ensures that there will be a `"help"` key in the map (in `ParsePayload`, which maps command/option to their corresponding parameters). However, I still think that we could return an empty vector here for two reasons:

- we get rid of panic
- an empty vector is properly handled later in the function

[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/command/help.rs:50]: > In Rust, we strive to avoid `unwrap()`s where possible. Static, syntactic guarantees are preferred (`match` and `if let` come in handy for these). To see why, keep reading.

Thank you for the thorough explanation! I definitely see how we benefit from the changes you introduced.

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:103]: I'm out of ideas. Maybe let's wrap the prefix in some kind of delimiters, e.g.
`Invalid 'replication' specification: 'foo==bar'`

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:105]: Once you get it, you'll even think in Iterators rather than any other (human or not) language ;)

File: src/bin/cql-stress-cassandra-stress/settings/option/node.rs [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    let nodes = handles
+        .nodes
+        .get()
+        .map(|list| list.split(',').map(|nd| nd.to_owned()).collect())
+        .unwrap();
+
+    let nodes = match file {
+        Some(f) => read_nodes_from_file(f)?,
+        None => nodes,
+    };
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/node.rs:117]: > We demand that a user provide a `nodes` param (because of the `unwrap()` at line 112)

The user doesn't need to provide it. We `unwrap` because this parameter has the default value of `localhost`.

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+        let map = self.arbitrary_params.as_mut().unwrap();
+        let key = split[0].to_string();
+        let val = split[1].to_string();
+        anyhow::ensure!(
+            !map.contains_key(&key),
+            "{} suboption has been specified more than once",
+            key
+        );
+        map.insert(key, val);
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:116]: > Is it a case that every multiparameter statically knows whether it accepts arbitrary parameters or not?

Yes, the `Param`s API objects are created internally, and are predefined (based on the parameters we support). So we can for sure program the so called "arbitrary parameters acceptance" using the Rust types (instead of boolean flag).

> My approach to the typesafe pattern here:

Thank you for presenting your approach. I understand the pattern and see the benefits. I have only one question. Let's take the `MultiParam::parse` function:

```rust
fn parse(&mut self, arg: &str) -> Result<()> {
        self.supplied_by_user = true;
        let arg_val = &arg[self.prefix.len()..];

        // Remove wrapping paranthesis.
        let arg_val = &arg_val[1..arg_val.len() - 1];

        // Iterate over comma-delimited sub-parameters.
        for subparam in arg_val.split(',') {
            // Check if the argument matches on of the predefined subparameters.
            match self.try_parse_predefined(subparam) {
                ParamMatchResult::Error(e) => return Err(e),
                ParamMatchResult::Match => continue,
                _ => (),
            }

            // If the argument didn't match any of the prefefined sub-parameters,
            // try to parse it as an arbitrary parameter (if applicable).
            anyhow::ensure!(
                self.accepts_arbitrary(),
                "Cannot accept parameter {}. {} command/option doesn't accept arbitrary parameters.",
                subparam,
                self.prefix
            );

            self.try_parse_arbitrary(subparam)?;
        }

        Ok(())
  }
``` 

From what I understood, in the snippet you provided, the `try_parse_arbitrary` is not implemented for `MultiParam<RejectsArbitratyParams>`. So now, we would technically need to implement `parse` method for both of the types separately (since it calls `try_parse_arbitrary` in one case, and doesn't call it in the other case).

In this particular case, the common `parse` logic is very simple, so there would be no problem copying this part of the code twice (or even moving it to some function implemented by `impl<T: ArbitraryParamAcceptance> MultiParam<T>`, and then calling the common logic from the actual implementations for the specific types). But what if the common logic would be much more complex?

But this is not the only solution. The other solution I see is to add `try_parse_arbitrary` method to the `ArbitraryParamsAcceptance` trait. In the case of `AcceptsArbitraryParams`, we would have the current implementation of `try_parse_arbitrary`, and in the case of `RejectsArbitraryParams` we would do nothing (nop), or maybe return the error (if for some reason the programmer didn't check if `accepts_arbitrary` before calling `try_parse_arbitrary).

What do you think about it?

[the remark this answers is not in the document]
Comment (wprzytula) [reply src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:116]: I've left this part unimplemented on purpose, for you to consider various possibilities :)
Your conclusions are in line with mine. Duplicating the logic is unacceptable; therefore, I'd go for the solution with extracting these lines:
```rust
    // If the argument didn't match any of the prefefined sub-parameters,
    // try to parse it as an arbitrary parameter (if applicable).
    anyhow::ensure!(
        self.accepts_arbitrary(),
        "Cannot accept parameter {}. {} command/option doesn't accept arbitrary parameters.",
        subparam,
        self.prefix
    );

    self.try_parse_arbitrary(subparam)?;
```
into a function whose implementation depends on `ArbitraryParamsAcceptance` type.
The implementation for `RejectArbitraryParams` would return the error:
```rust
    anyhow::anyhow!("Cannot accept parameter {}. {} command/option doesn't accept arbitrary parameters.",
                subparam,
                self.prefix
    )
```
and the one for `AcceptArbitraryArguments` would just call `try_parse_arbitrary()`.

The idea is, once again, that we want to make it **IMPOSSIBLE** to make an error, so we want the programmer to not have to call `accepts_arbitrary()` anytime here.

File: src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs [hunk 15]
```diff
@@ -10,6 +10,74 @@ lazy_static! {
...
+/// Multiparameters may or may not accept arbitrary parameters.
+/// That's why we introduce the trait responsible for accepting such parameters.
+/// [MultiParam] is generic over the types that implement this trait.
+/// See [AcceptsArbitraryParams] and [RejectsArbitraryParams].
+pub trait ArbitraryParamsAcceptance: Sized + Default {
+    fn accepts_arbitrary(&self) -> bool;
+    fn try_parse_arbitrary(&mut self, param_name: &str, arg: &str) -> Result<()>;
+}
+
+/// [MultiParam<AcceptsArbitratyParams>] will accept all of the arbitrary parameters.
```
Comment (wprzytula) [inline src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs:22]: `Arbitraty` -> `Arbitrary`

**Review by wprzytula — COMMENTED** [review]
I got confused during re-review, because the fixes were done in separate commits and put at the end. This resulted in me writing comments for not-yet-resolved issues, which were actually resolved earlier.
Please reposition the commits with fixes so that:
1) fixes to things NOT introduced in this PR come first,
2) fixes to things introduced in this PR either:
    - come immediately after the commit introducing the previous version (this is recommeded if the fixup is controversial and is likely to be contested by reviewers), or
    - get squashed as a fixup into the commit introducing the previous version (recommended for noncontroversial fixups).

File: src/bin/cql-stress-cassandra-stress/settings/option/node.rs [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+    fn from_handles(handles: NodeParamHandles) -> Result<NodeOption> {
+        let datacenter = handles.datacenter.get();
+        let whitelist = handles.whitelist.supplied_by_user();
+        let file = handles.file.get();
+        let nodes = handles
+            .nodes
+            .get()
+            .map(|list| list.split(',').map(|nd| nd.to_owned()).collect())
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/option/node.rs]: We get rid of this `unwrap` in this commit: https://github.com/scylladb/cql-stress/commit/e624db2b5ee331d504762c0da8fd6d608f33df4a.

**Review by wprzytula — APPROVED** [review]
Please remember to squash you fixup commits after review is finished.

**Comment by muzarski** [comment]
v4:
- Squashed fixup commits.

**Comment by muzarski** [comment]
v4:
- Squashed fixup commits

**Comment by muzarski** [comment]
v5:
- rebased on top of master

File: src/bin/cql-stress-cassandra-stress/settings/option/mod.rs [hunk 17]
```diff
@@ -0,0 +1,24 @@
+use anyhow::Result;
+
+pub struct Options();
```
Comment (piodul) [inline src/bin/cql-stress-cassandra-stress/settings/option/mod.rs:11]: Nit: this also works and looks nicer:

```rust
pub struct Options;
```

**Review by piodul — COMMENTED** [review]
Some nitpicks, otherwise LGTM

File: src/bin/cql-stress-cassandra-stress/settings/mod.rs [hunk 18]
```diff
@@ -142,10 +164,32 @@ where
...
+                .collect::<Vec<_>>();
+            unknowns.join("\n")
+        };
+
+        // Ensure that all of the CLI arguments were consumed.
+        // If not, then unknown arguments appeared so we return the error.
+        anyhow::ensure!(
+            payload.is_empty(),
+            "Error processing CLI arguments. The following were ignored:\n{}",
+            build_unknown_arguments_err_message()
```
[the remark this answers is not in the document]
Comment (muzarski) [reply src/bin/cql-stress-cassandra-stress/settings/mod.rs:172]: The expansion of `anyhow::ensure` macro in this case:
```rust
// Recursive expansion of ensure! macro
// =====================================

if !(payload.is_empty()) {
    return $crate::__private::Err($crate::Error::msg({
        let res = $crate::fmt::format(::core::fmt::Arguments::new_v1(
            &["Error processing CLI arguments. The following were ignored:\n"],
            &[::core::fmt::ArgumentV1::new(
                &(build_unknown_arguments_err_message()),
                ::core::fmt::Display::fmt,
            )],
        ));
        res
    }));
}
```

Clearly, the error message is built lazily, only when the check fails.

**Review by piodul — APPROVED** [review]
